### DISCUSSION TITLE
blog/ref-callbacks-react-19-and-the-compiler

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#113] [Announcements]
# blog/ref-callbacks-react-19-and-the-compiler

Reflecting on useCallback, how the Compiler changed my thinking about it, and what React 19 has in store for Ref Callbacks...

https://tkdodo.eu/blog/ref-callbacks-react-19-and-the-compiler#inject-comments

**Comment by 3bdallahz** [comment]
in the last example, doesn't the callback(and cleanup) get called on every render
and that would be creating ResizeObserver and cleaning it up on every render

Should we wrap it with a useCallback/useEvent?

**Reply by TkDodo** [reply]
Yes, that is what's happening. We can add `useCallback` as a performance optimization, but our code will work with or without it. I mentioned that here:

> A [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) can address both problems, and we can now create one and clean it up inside our ref callback (with or without useCallback)

**Reply by 3bdallahz** [reply]
unrelated question, but would be nice to know
how should we merge/compose refs that return cleanups
I don't think merge-refs or @radix-ui/react-compose-refs handle this

**Reply by TkDodo** [reply]
They can handle, as they would otherwise fail with functions, which existed forever:

https://github.com/wojtekmaj/merge-refs/blob/0a964c5ace86c902c89bba1430e2be3227dfa9cd/src/index.ts#L28-L32

But also, if you disregard other functions, merging ref just makes sure that `ref.current` of each of those refs is assigned the node. You can also do this manually. Suppose there’s a `ref` already (which is an object), then you can do:

```
function Component({ ref }) {

  const myRef = useCallback((node) => {
    ref.current = node

    // do whatever I want
  }, [])

  return <div ref={myRef} />
}
```

**Reply by 3bdallahz** [reply]
```ts
      if (typeof inputRef === 'function') {
        inputRef(ref);
      } else if (inputRef) {
        (inputRef as React.MutableRefObject<T | null>).current = ref;
      }
```

but cleanup that is possibly retruned from `inputRef` is not handled
but I think just assigning it manually like you suggested is actually simpler
Thanks @TkDodo

**Reply by TkDodo** [reply]
oh yeah that’s a good point, I guess they’d need to somehow add support for that. Not sure if chaining ref callbacks is a good idea - maybe we should stick to ref objects for that 🤔

**Reply by chaance** [reply]
~~We'll need to update Radix, but it's a relatively simple change if you're doing something similar in your own code.~~ Updated and released a new version yesterday to support cleanup functions (thanks @dmitri-gb!). [Here's the code](https://github.com/radix-ui/primitives/blob/dac4fd8ab0c1974020e316c865db258ab10d2279/packages/react/compose-refs/src/composeRefs.tsx) if you're doing something similar in your own app.

**Reply by 3bdallahz** [reply]
Don't we need to also set non-function refs to null in cleanup, or that is not necessary?

**Reply by chaance** [reply]
@3bdallahz Yep, good callout. Didn't realize that the ref callback itself is no longer called when the element is detached when a cleanup as returned (but it makes total sense that it wouldn't!).

**Reply by KurtGokhan** [reply]
@chaance there is a slight edge case where your code won't set null to the old refs if none of the refs return a callback cleanup. (`if(hasCleanup)` is unnecessary imo).

[Here](https://gist.github.com/KurtGokhan/9aafd8e83c9bc6a2946fe2dc7f2c1d19) is another implementation I wrote a while ago.

Edit: Nevermind. For it to work with both React 19 and 18, your code is correct.

**Comment by f-z-coder** [comment]
As you said ref callback run on mount and very re-render without use callback,Does cleanup function run on every re-render. Or it just run while unmount node ??

**Reply by TkDodo** [reply]
yes, the cleanup function runs every time before the _new_ ref function would run. So if it’s an inline function, it also runs on every render for a proper clean-up.

**Comment by Efimenko** [comment]
Hi Dominik, thank you for continuing this great series of articles.
I have one question regarding your examples. So, instead of using `useCallback`, you've moved `scrollIntoView` outside the component. But what if the ref callback depends on data that is available only in render? For instance, the same example as you have, but instead of an immediate scroll on node mount, it checks that the `id` of the node is the same as you have in `hash`:
```ts
const useScrollToSectionByHash = () => {
  const { hash: hashSectionId } = useLocation();

  const sectionIdFromHash = hashSectionId.slice(1);

  return (node: null | HTMLDivElement) => {
    if (sectionIdFromHash === node?.id) {
      node.scrollIntoView();
    }
  };
};
```
So, here is the question: What approach do you think is appropriate here? It seems like the only possible way is to use `useCallback`, because any re-render will re-scroll you back to the node if you have `hash`, but at the same time, `useCallback` semantically should not be used for such a case.
P.S. I know that the browser can automatically scroll to the node with the `id` if you have the same hash, but let's say that the node is lazy loaded and unavailable immediately.

**Reply by TkDodo** [reply]
I left out this example because I couldn’t find a good one where we actually need to depend on a prop. Sathya posted a better workaround than `useCallback` for this [on bsky](https://bsky.app/profile/gsathya.bsky.social/post/3lbjz4ffvhs2t):

```
const idRef = useRef(null);
const scroller = (node) => {
  if (idRef.current === null || idRef.current !== id) {
    node?.scrollIntoView({ behavior: "smooth" });
    idRef.current = id;
  }
};
```

But tbh, this looks more complicated than just having a `useLayoutEffect` if you really need that.

**Reply by 832bb9** [reply]
I feel it requires additional mental effort from developer to think about different approaches: 
- Should I extract this callback to the global scope?
- Should I use useLayoutEffect?
- Should I use ref and manage it's value in callback?

It would be much easier if there was something like useEffect that runs only when dependencies change, but also can be used as ref callback. React docs doesn’t suggest replacing callback refs with effects, but it doesn't explain how to avoid extra calls on re-renders. I hope they’ll eventually add something similar to https://react.dev/reference/react/experimental_useEffectEvent for callback refs. Until then, I’m using my own custom hook, which ensures callback refs aren’t called until dependencies change, while still following React’s purity rules.

```typescript
// https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents
export const useInit = <Return extends unknown>(init: () => Return) => {
  const ref = useRef<null | Return>(null);

  ref.current ??= init();

  return ref.current;
};

const createCache = () => new Map();

const mapDepsToCacheKey = (deps: DependencyList) => JSON.stringify(deps);

const useCallbackRef = <Element extends unknown>(
  callback: RefCallback<Element>,
  deps: DependencyList
) => {
  const cache = useInit(() => createCache());

  const key = useMemo(() => mapDepsToCacheKey(deps), deps);

  if (!cache.has(key)) {
    cache.set(key, callback);
  }

  return cache.get(key) as typeof callback;
};
```

Modifying refs during render is not allowed, but we can use refs to store a cache and get/set values based on a key. If you only need to support serializable dependencies (like an 'id' string), a simple Map and JSON.stringify can be used.

This utility hook can be used as follows:

```typescript
const scrollToSection = useCallbackRef(
  (node: null | HTMLDivElement) => {
    if (!node || !node.id || node.id !== sectionId) {
      return;
    }

    node.scrollIntoView({ behavior: "smooth" });
  },
  [sectionId]
);
```

Here's a sandbox with the full example https://codesandbox.io/p/sandbox/hxvn57

**Comment by nimeshnayaju** [comment]
Hi @TkDodo , thank you for the follow-up article—it was really insightful! In the example you shared, would it be more appropriate to use `useSyncExternalStore` since you’re subscribing to an external store? I understand the intent of the example, but it makes me wonder whether it’s always worth defaulting to `useLayoutEffect` and/or `useEffect` for most (if not all) effect related logic and use `useRef` to obtain the ref to the DOM node.

```tsx
const measuredRef = (node) => {
  const observer = new ResizeObserver(([entry]) => {
    setHeight(entry.contentRect.height)
  })

  observer.observe(node)

  return () => {
    observer.disconnect()
  }
}
```

**Reply by nimeshnayaju** [reply]
I think what makes sense to me is to combine `useRef` and `useCallback` to, for instance, manage a list of refs to DOM nodes.

```tsx
const refs = useRef(new Map<string, HTMLElement>())

...

<li
  key={id}
  ref={node => {
    refs.current.set(id, node);
    return () => {
      refs.current.delete(id);
    };
  }}
>
)
```

**Reply by TkDodo** [reply]
`useSyncExternalStore` works great if you subscribe to primitive values. It’s not great with setting up observers

**Comment by deanolium** [comment]
One issue I could see with this is basically the main use case of why you measure your DOM elements. That is, to use these measurements to alter a component in some way. For instance, maybe you have a caption box showing the dimensions of a child component?

Using the ref callback along with a useState would surely mean you get the measurement when the component ref callback is called, which will then trigger the useState causing a re-render. But the gotcha here is that the first render, before the state is set to the measured value, will be put on screen. It would be quickly replaced since most renders cycles are very fast, but it would still look a bit glitchy depending on what you're doing with the measured value.

Whilst useLayoutEffect blocks the browser from repainting the screen whilst it's running, and if the useLayoutEffect callback causes a re-render, that will then happen before the browser is allowed to repaint. This means the user won't see that first render that is measured if the state has to change. This gives a much better user experience.

That said, measuring elements in order to then do something visually with the measurement is really the only time that useLayoutEffect is the way to go. For dynamically resizeable elements, this is less important since the one 'frame' of measurement lag is much more accepted after that initial render.

**Reply by TkDodo** [reply]
to the best of my knowledge, ref callbacks have the same timing as useLayoutEffect. Both happen before the browser paints.
